dataservice/api/participant/models.py

```diff
@@ -27,3 +29,39 @@ class Participant(db.Model, Base):
     demographic = db.relationship(Demographic, backref='participant',
                                   uselist=False, cascade="all, delete-orphan",
                                   lazy=True)
+
+    def upstream_immediate_relatives(self):
+        """
+        Convenience method to get immediate related participants upstream of
+        self.
+
+        An upstream relative is the participant pointing to self in the family
+        relationship.
+        For example for the family relationship: P1 --> P2 (where P2 is self),
+        P1 is the upstream relative of P2.
+        """
+        for r in self.incoming_family_relationships:
+            yield r.participant
+
+    def downstream_immediate_relatives(self):
+        """
+        Convenience method to get related immediate relatives downstream from
+        self.
+
+        A downstream relative is the participant that self points to in the
+        family relationship.
+        For example for the family relationship: P1 --> P2 (where P1 is self),
+        P2 is the downstream relative of P1.
+        """
+        for r in self.outgoing_family_relationships:
+            yield r.relative
+
+    def immediate_relatives(self):
```

Comment:
Could this be a class property?

Reply:
I think you're asking whether `immediate_relatives` could be an instance variable (list of Participant objects)? If so, I think this is typically done using association proxy, but I'm not quite sure how to implement it for a self-referential many to many. Even for a normal many-to-many, I'm still trying to figure out how to get association proxy to work. I tried using association proxy on the Study-Participant many-to-many and had issues so I abandoned it for now...

Reply:
I meant you could use `@property` here so it looks like a property rather than a function.

Reply:
Ah gotcha. Well this is exactly what association proxy is for and i figured out how to use it for the other many to many relationships yayyyy. So I'll see if i can get it to work with this one as well. If not then maybe I can use the property decorator as a temporary workaround

Reply:
Hm, do you mean having this field be a relationship with secondary joins?